src/main/java/org/opensearch/knn/index/query/parser/MethodParametersParser.java

```diff
@@ -0,0 +1,100 @@
+import org.opensearch.knn.index.query.model.HNSWAlgoQueryParameters;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER;
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_EF_SEARCH;
+import static org.opensearch.knn.index.IndexUtil.isClusterOnOrAfterMinRequiredVersion;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.EF_SEARCH_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.NAME;
+import static org.opensearch.knn.index.query.parser.KNNXParserUtil.parseJsonObject;
+
+@Getter
+public class MethodParametersParser {
+
+    private static final Set<String> VALID_METHOD_PARAMETERS = ImmutableSet.of(METHOD_PARAMETER_EF_SEARCH);
+
+    @SneakyThrows(IOException.class)
+    public static AlgoQueryParameters streamInMethodParameters(StreamInput in) {
+        Integer efSearch = null;
+        if (isClusterOnOrAfterMinRequiredVersion(METHOD_PARAMETER_EF_SEARCH)) {
+            efSearch = in.readOptionalInt();
+        }
+        return efSearch != null ? HNSWAlgoQueryParameters.builder().efSearch(efSearch).build() : null;
+    }
+
+    @SneakyThrows(IOException.class)
+    public static void streamOutMethodParameters(final StreamOutput out, final AlgoQueryParameters algoQueryParameters) {
+        final Optional<HNSWAlgoQueryParameters> hnswMethodParameters = extractHnswAlgoParameters(algoQueryParameters);
+
+        if (isClusterOnOrAfterMinRequiredVersion(METHOD_PARAMETER_EF_SEARCH)) {
+            // Write false even if it doesn't cast to hnsw, so we can deserialize w/o ambiguity as parameters get added
+            out.writeOptionalInt(hnswMethodParameters.flatMap(HNSWAlgoQueryParameters::getEfSearch).orElse(null));
+        }
+    }
+
+    public static Optional<HNSWAlgoQueryParameters> extractHnswAlgoParameters(final AlgoQueryParameters algoQueryParameters) {
+        return Optional.ofNullable(algoQueryParameters)
+            .filter(HNSWAlgoQueryParameters.class::isInstance)
+            .map(HNSWAlgoQueryParameters.class::cast);
+    }
+
+    @SneakyThrows(IOException.class)
+    public static AlgoQueryParameters fromXContent(final XContentParser parser) {
+        final Map<String, Object> parameters = parseJsonObject(parser);
+        if (parameters.isEmpty()) {
+            throw new IllegalArgumentException("[" + NAME + "] method_parameter cannot be empty");
+        }
+
+        for (String jsonkey : parameters.keySet()) {
+            if (!VALID_METHOD_PARAMETERS.contains(jsonkey)) {
+                throw new IllegalArgumentException("[" + NAME + "] unknown parameter " + jsonkey + " found.");
+            }
+        }
+
+        return Optional.ofNullable((Integer) parameters.get(METHOD_PARAMETER_EF_SEARCH))
+            .filter(ef -> EF_SEARCH_FIELD.match(METHOD_PARAMETER_EF_SEARCH, parser.getDeprecationHandler()))
```

Comment:
should we have some checks to ensure that ef_Search is a +ve value. my worry is if some once gives a value which is out of int range we typecasting it to int will lead to negative value. or someone my pass -ve values. So lets ensure we are throwing right exceptions in those cases.

Reply:
> should we have some checks to ensure that ef_Search is a +ve value

+ve check is present [here](https://github.com/opensearch-project/k-NN/pull/1707/files#diff-6d0df17543181185170f65c116cdb3514d910e1557f012b7959611ae92678080R222). Overall I would want to decouple parsing and validations moving forward and I have tried to reflect that with ef_search